pdfminer/pdfdocument.py

```diff
@@ -348,7 +348,9 @@ def compute_encryption_key(self, password):
         password = (password + self.PASSWORD_PADDING)[:32]  # 1
         hash = md5.md5(password)  # 2
         hash.update(self.o)  # 3
-        hash.update(struct.pack('<L', self.p))  # 4
+        # See https://github.com/pdfminer/pdfminer.six/issues/186
+        unsigned_p = self.p if self.p > 0 else self.p + (1 << 32)
```

Comment:
This should be done directly after `self.p` is created. E.g. after or on line 310. For example by swapping the `int_value()` for a new function `uint_value()`.

Reply:
I think it should not, as it would break `self.is_printable`, `self.is_modifiable` and `self.is_extractable`

Reply:
Yes, I believe that those variables currently have wrong values. 

As an example from the PDF reference: 

> For example, assuming revision 2 of the security handler, the value -44 permits printing and copying but disallows modifying the contents and annotations.

With the current code, this is directly interpreted as a bit string (see below). The third bit (printing) is correctly one, but the fifth bit (copying) is wrongly zero, and the fourth bit (modifying) is wrongly one. 

```
>>> '{:032b}'.format(-44)
'-0000000000000000000000000101100'
```

With your new code, this creates another bit string (see below). 

```
>>> p = -44
>>> unsigned_p = p if p > 0 else p + (1 << 32)
>>> '{:032b}'.format(unsigned_p)
'11111111111111111111111111010100'
```

With this new code, the third bit (printing) is correctly one, the fifth bit (copying) is correctly one, and the fourth bit (modifying) is correctly zero. 

Also note that all the higher-order bits are one with the new code. This is also an improvement. 

My conclusion: your code should be used to interpret the value of `P`.

Reply:
Note: changing this also influences #350.

Reply:
Actually it's the same, as for the bit level `&` python considers classical two's complement, so it doesn't matter where we do it, but sadly it does not influence #350 :(

```for i in range(1000000):
    p = -i
    up = p+(1<<32)
    assert (p & 4) == (up & 4)
    assert (p & 8) == (up & 8)
    assert (p & 16) == (up & 16)```

Reply:
Wow, that's a nice property! It took some time for me to wrap my head around this but I think I got it. 

I think it would still be preferable to do this operation in a function, just after reading the value. This makes it a whole lot easier to understand what is going on, and look more like the PDF Reference. 

Do you have the time to work on this? Or do you want me to have a go, and you doing the review?

Reply:
Do it how you like it, since you're going to maintain it and it's just a matter of choosing if self.p (and maybe the other int values?) should follow the internal representation (signed) or the one used for packing (unsigned)